javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,293 @@
+      }
+    }
+    removeNext(fromIndex, listIterator(fromIndex))
+  }
+
+  def set(index: Int, element: E): E =
+    throw new UnsupportedOperationException
+
+  def subList(fromIndex: Int, toIndex: Int): List[E] = {
+    if (fromIndex < 0 || toIndex > size)
+      throw new IndexOutOfBoundsException
+    else if (fromIndex > toIndex)
+      throw new IllegalArgumentException
+
+    self match {
+      case _: RandomAccess =>
+        new AbstractListView(fromIndex, toIndex) with RandomAccess {
+          override def listIterator(index: Int): java.util.ListIterator[E] = {
+            checkIndexOnBoundsOfView(index)
+            // Iterator that accesses the original list directly
+            new RandomAccessListIterator(fromIndex + index, fromIndex, toIndex)
+          }
+        }
+      case _ =>
+        new AbstractListView(fromIndex, toIndex) {
+          override def listIterator(index: Int): java.util.ListIterator[E] = {
+            checkIndexOnBoundsOfView(index)
+            // Iterator that accesses the original list using it's iterator
+            new BackedUpListIterator(index, fromIndex)
+          }
+        }
+    }
+  }
+
+  private def checkIndexOnBounds(index: Int): Unit = {
+    if (index < 0 || index > size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  private abstract class AbstractListView(fromIndex: Int,
+                                          private var toIndex: Int) extends AbstractList[E] {
+
+    override def add(e: E): Boolean = {
+      add(size, e)
+      toIndex += 1
+      true
+    }
+
+    override def add(index: Int, e: E): Unit = {
+      checkIndexOnBoundsOfView(index)
+      self.add(fromIndex + index, e)
+    }
+
+    override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+      checkIndexOnBoundsOfView(index)
+      val success = self.addAll(fromIndex + index, c)
+      if (success)
+        toIndex += c.size
+      success
+    }
+
+    override def addAll(c: Collection[_ <: E]): Boolean =
+      addAll(size, c)
+
+    override def clear(): Unit = {
+      self.removeRange(fromIndex, toIndex)
+      toIndex = fromIndex
+    }
+
+    def get(index: Int): E = {
+      checkIndexInBoundsOfView(index)
+      self.get(fromIndex + index)
+    }
+
+    override def remove(index: Int): E = {
+      checkIndexInBoundsOfView(index)
+      val elem = self.remove(fromIndex + index)
+      toIndex -= 1
+      elem
+    }
+
+    override def set(index: Int, e: E): E = {
+      checkIndexInBoundsOfView(index)
+      self.set(fromIndex + index, e)
+    }
+
+    def size(): Int = toIndex - fromIndex
+
+    protected def checkIndexInBoundsOfView(index: Int): Unit = {
+      if (index < 0 || index >= size)
+        throw new IndexOutOfBoundsException(index.toString)
+    }
+
+    protected def checkIndexOnBoundsOfView(index: Int): Unit = {
+      if (index < 0 || index > size)
+        throw new IndexOutOfBoundsException(index.toString)
+    }
```

Comment:
I would suggest that these two methods be defined once and for all as `protected[this]` in `AbstractList`. So that they don't have to be repeated in every view/class/etc.